lfs/api.go

```diff
@@ -0,0 +1,93 @@
+package lfs
+
+import (
+	"io"
+	"sync"
+)
+
+// An abstract interface providing state & resource management for a specific Endpoint across
+// potentially multiple requests
+type ApiContext interface {
+	// Get the endpoint this context was constructed from
+	Endpoint() Endpoint
+	// Close the context & any resources it's using
+	Close() error
+
+	// Download a single object, return reader for data, size and any error
+	// Essentially the same as calling DownloadCheck() then DownloadObject()
+	Download(oid string) (io.ReadCloser, int64, *WrappedError)
+	// Check whether an object is available for download and return an object resource if so
+	DownloadCheck(oid string) (*ObjectResource, *WrappedError)
+	// Download a single object from an already identified resource from DownloadCheck(), return reader for data, size and any error
+	DownloadObject(obj *ObjectResource) (io.ReadCloser, int64, *WrappedError)
+	// Check whether an upload would be accepted for an object and return the resource to use if so
+	UploadCheck(oid string, sz int64) (*ObjectResource, *WrappedError)
+	// Perform the actual upload of an object having identified it will be accepted and the resource to use
+	UploadObject(o *ObjectResource, reader io.Reader) *WrappedError
+	// Perform a batch request for a number of objects to determine what can be uploaded/downloaded
+	Batch(objects []*ObjectResource) ([]*ObjectResource, *WrappedError)
```

Comment:
Yeah the concept that the context can survive over multiple API calls is necessary until you can guarantee that only one API call will be made per command, unless you want a separate SSH connection to be fired up for every file. Even after the *All() batch methods are implemented though, I think it's nice to have that flexibility - to replace it would just be equivalent to a global context per API and I'm not sure that actually simplifies much; instead baking in an assumption.